mesos-client/src/main/scala/mesosphere/mesos/client/MesosClient.scala

```diff
@@ -0,0 +1,514 @@
+
+    val subscribedHandler: Flow[Event, Event, NotUsed] = Flow[Event].map { event =>
+      if (event.subscribed.isDefined) {
+        val ctx = context.updateAndGet(c => c.copy(frameworkId = Some(event.subscribed.get.frameworkId)))
+        // We save `frameworkId` in the context and successfully complete the promise, meaning the calls in sink
+        contextPromise.success(ctx)
+      }
+      event
+    }
+
+    def connectionSource(host: String, port: Int) =
+      Source.single(request)
+      .via(log(s"Connecting to the new leader: $host:$port"))
+      .via(httpConnection(host, port))
+      .via(log("HttpResponse: "))
+      .via(connectionHandler)
+
+    connectionSource(context.get().host, context.get().port)
+      .recoverWithRetries(conf.redirectRetires(), {
+        case MesosRedicrectException(_) => connectionSource(context.get().host, context.get().port)
+      })
+      .idleTimeout(conf.idleTimeout().seconds)
+      .buffer(conf.sourceBufferSize(), overflowStrategy)
+      .via(dataBytesExtractor)
+      .via(recordIoScanner)
+      .via(eventDeserializer)
+      .via(subscribedHandler)
+      .via(log("Received mesos Event: "))
+  }
+
+  override lazy val (killSwitch: UniqueKillSwitch, mesosSource: Source[Event, NotUsed]) = source
+    .viaMat(KillSwitches.single)(Keep.right)
+    .toMat(BroadcastHub.sink)(Keep.both)
+    .run()
+
+
+  /**
+  Mesos sink is an akka-stream `Sink[Call, Notused]` that sends calls to mesos. Every call is send via a new (and pooled) connection.
+  The flow visualized:
+
+       |  |  |
+       v  v  v
+      ----------
+     | MergeHub | (1)
+      ----------
+          |
+          v
+     ------------
+    | Call       |
+    | Enhancer   | (2)
+     ------------
+          |
+          v
+     ------------
+    | Event      |
+    | Serializer | (3)
+     ------------
+          |
+          v
+     ------------
+    | Request    |
+    | Builder    | (4)
+     ------------
+          |
+          v
+     ------------
+    | Http Sink  | (5)
+     ------------
+
+    1. MergeHub allows dynamic "fan-in" junction point for mesos calls from multiple producers.
+    2. Call Enhancer updates the call with the framework Id from the connection context
+    3. Event Serializer serializes calls to byte array
+    4. Build a HTTP request from the data using `mesosStreamId` header from the context
+    5. Http Sink creates a new connection using akka's `Http().singleRequest` and sends the data
+
+    Note: Merge hub will wait for the _connection context_ object to be fully initialized first meaning that we have current leader's
+    `host`, `port` and `Mesos-Stream-Id` to send the events to.
+    */
+  private val sink: Sink[(ConnectionContext, HttpRequest), Future[Done]] = Sink.foreach[(ConnectionContext, HttpRequest)]{
+    case (_, request) =>
+      Http().singleRequest(request).map { response =>
```

Comment:
It's extremely dangerous to use Sink.foreach here. It will fire http requests with a light-speed rate, quickly overflowing a small http pool (the pool is shared for the entire actor system) and will possible cause unexpected failures. I had this in production and it's definitely not something you'd like to have.

Much better solution is to use a Flow-based variant of the http pool

Reply:
Very good point. I played with `Http().cachedHostConnectionPool` and `Http().superPool()` but it turns out that they can't guarantee delivery of the messages in the same order they were sent. I'm using `Http().outgoingConnection` that holds one long-living connection to mesos (probably an improvement by itself) and sends calls over it. I also left an explaining comment in code.